x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java

```diff
@@ -0,0 +1,338 @@
+            double metricValue, List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) {
+        this(name, sortFormat, sortOrder, sortValueFor(sortValue), metricName, metricValue, pipelineAggregators, metaData);
+    }
+
+    static InternalTopMetrics buildEmptyAggregation(String name, String metricField,
+            List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) {
+        return new InternalTopMetrics(name, DocValueFormat.RAW, SortOrder.ASC, null, metricField, Double.NaN, pipelineAggregators,
+                metaData);
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public InternalTopMetrics(StreamInput in) throws IOException {
+        super(in);
+        sortFormat = in.readNamedWriteable(DocValueFormat.class);
+        sortOrder = SortOrder.readFromStream(in);
+        sortValue = in.readOptionalNamedWriteable(SortValue.class);
+        metricName = in.readString();
+        metricValue = in.readDouble();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(sortFormat);
+        sortOrder.writeTo(out);
+        out.writeOptionalNamedWriteable(sortValue);
+        out.writeString(metricName);
+        out.writeDouble(metricValue);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return TopMetricsAggregationBuilder.NAME;
+    }
+
+    @Override
+    public Object getProperty(List<String> path) {
+        if (path.isEmpty()) {
+            return this;
+        }
+        if (path.size() == 1 && metricName.contentEquals(path.get(1))) {
+            return metricValue;
+        }
+        throw new IllegalArgumentException("path not supported for [" + getName() + "]: " + path);
+    }
+
+    @Override
+    public InternalTopMetrics reduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        Iterator<InternalAggregation> itr = aggregations.iterator();
+        InternalTopMetrics first;
+        do {
+            if (false == itr.hasNext()) {
+                // All of the aggregations are empty.
+                return buildEmptyAggregation(name, metricName, pipelineAggregators(), getMetaData());
+            }
+            first = (InternalTopMetrics) itr.next();
+        } while (first.sortValue == null);
+        DocValueFormat bestSortFormat = first.sortFormat;
+        SortValue bestSortValue = first.sortValue;
+        double bestMetricValue = first.metricValue;
+        int reverseMul = first.sortOrder.reverseMul();
+        while (itr.hasNext()) {
+            InternalTopMetrics result = (InternalTopMetrics) itr.next();
+            if (result.sortValue == null) {
+                // Don't bother checking empty results.
+                continue;
+            }
+            if (reverseMul * bestSortValue.compareTo(result.sortValue) > 0) {
```

Comment:
Just for my understanding/clarification, the way the SortValues are setup is basically:

1. If the types differ (`Long` compared to `Double`, it does a comparison on the class name and uses that for ordering)
2. Otherwise, if the types are the same it does the type-specific comparison

That correct?

Reply:
It compares the *writeable* name for the sorting, not the class name. But you had it pretty much. This sort of comparison will only be done on the reducing node so I don't *think* it is super important what key it uses, just that it uses *something*. And the writeable name is right there.

Though maybe it does matter if we reduce in multiple phases on multiple nodes. Is that a thing that cross cluster search does? If so I should probably be a little more explicit here.

Reply:
:+1: 

I believe CCS can potentially execute like that.  One mode streams all shard results (from all nodes/clusters) to the coordinator and reduces.  But a different mode does an incremental reduce in a per-cluster coordinator and streams those "intermediate" shard results to the original coordinator for a final reduce.  I forget the name of that mode but it's basically designed for higher-latency CCS where you don't want to ship all the shard results across the world.

But I think writeable name should be fine even in that case, right?  It just needs to be consistent?

Reply:
It *does* need to be consistent. But I think it is important to add tests and comments explaining *why* it needs to be consistent. I guess writeable name has to be consistent for the wire protocol to work at all.